storage_broker/src/bin/storage_broker.rs

```diff
@@ -662,52 +667,70 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
-                            res_resp.map(|resp| resp.map(EitherBody::Right))
-                        }
-                    }
-                }))
+    let tcp_listener = TcpListener::bind(&args.listen_addr).await?;
+    info!("listening on {}", &args.listen_addr);
+    loop {
+        let (stream, addr) = match tcp_listener.accept().await {
+            Ok(v) => v,
+            Err(e) => {
+                info!("couldn't accept connection: {e}");
+                continue;
             }
-        }))
-        .await?;
-    Ok(())
+        };
+
+        let mut builder = hyper_util::server::conn::auto::Builder::new(TokioExecutor::new());
+        builder
+            .http2()
+            .keep_alive_interval(Some(args.http2_keepalive_interval));
+
+        let storage_broker_server_cloned = storage_broker_server.clone();
+        let connect_info = stream.connect_info();
+        let service_fn_ = async move {
+            service_fn(move |mut req| {
+                // That's what tonic's MakeSvc.call does to pass conninfo to
+                // the request handler (and where its request.remote_addr()
+                // expects it to find).
+                req.extensions_mut().insert(connect_info.clone());
+
+                // Technically this second clone is not needed, but consume
+                // by async block is apparently unavoidable. BTW, error
+                // message is enigmatic, see
+                // https://github.com/rust-lang/rust/issues/68119
+                //
+                // We could get away without async block at all, but then we
+                // need to resort to futures::Either to merge the result,
+                // which doesn't caress an eye as well.
+                let mut storage_broker_server_svc = storage_broker_server_cloned.clone();
+                async move {
+                    if req.headers().get("content-type").map(|x| x.as_bytes())
+                        == Some(b"application/grpc")
+                    {
+                        let res_resp = storage_broker_server_svc.call(req).await;
+                        // Grpc and http1 handlers have slightly different
+                        // Response types: it is UnsyncBoxBody for the
+                        // former one (not sure why) and plain hyper::Body
+                        // for the latter. Both implement HttpBody though,
+                        // and EitherBody is used to merge them.
+                        res_resp.map(|resp| resp.map(EitherBody::Left))
+                    } else {
+                        let res_resp = http1_handler(req).await;
+                        res_resp.map(|resp| resp.map(EitherBody::Right))
+                    }
+                }
+            })
+        }
+        .await;
+
+        let res = builder
+            .serve_connection(TokioIo::new(stream), service_fn_)
+            .await;
+
+        if let Err(e) = res {
+            info!("error serving connection from {addr}: {e}");
```

Comment:
is there a defined error case for this if we have pre-bound the address?

Reply:
note I think this is called from main so I don't think it can be a tcp timeout or anything scoped to that level.

Reply:
It's trivially triggerable. E.g. this can cause an `hyper::Error(Parse(UriTooLong))` error:
```
printf "http://127.0.0.1:3000/%0100000d" "9" | xargs curl
```
Tried it out with the hello example from the [hyper repo](https://github.com/hyperium/hyper/)

Reply:
it doesn't, I've added a manual spawn.

FTR [this](https://github.com/hyperium/hyper/blob/bb51c81b74cbbeaa922d52d4472b25eaf3c62eff/src/proto/h1/role.rs#L436-L447) is a list of possible errors for HTTP1: